src/trusted/service_runtime/nacl_syscall_common.c

```diff
@@ -4933,51 +4939,45 @@ int32_t NaClSysWaitpid(struct NaClAppThread *natp,
     NaClRemoveZombie(nap, pid);
 
     ret = pid;
-    *stat_loc_ptr = nap_child->exit_status;
+    if (stat_loc_ptr) *stat_loc_ptr = nap_child->exit_status;
     goto out;
   }
 
   // if WAIT_ANY, we'll busy loop on all children, except in the case of WNOHANG where we only loop once
   if (pid <= 0) {
     while(1){
 
-      /* Cycle through possible cages up to the fork number (max amount of created cages) */
-      for (int cage_id = 0; cage_id < fork_num; cage_id++) {
+      /* make sure children exist, if not send ABI_ECHILD */
+      if (!nap->num_children && !nap->zombies.num_entries) {
+        ret = -NACL_ABI_ECHILD;
+        NaClXCondVarBroadcast(&nap->children_cv);
+        NaClXMutexUnlock(&nap->children_mu);
+        goto out;
+      }
+
+      NaClXMutexLock(&nap->children_mu);
+      NaClLog(1, "Thread children count: %d\n", nap->num_children);
 
-        NaClXMutexLock(&nap->children_mu);
+      // wait here until a signal is sent to check 
+      NaClXCondVarTimedWaitRelative(&nap->children_cv, &nap->children_mu, &timeout);
 
-        /* make sure children exist, if not send ABI_ECHILD */
-        if (!nap->num_children) {
-          ret = -NACL_ABI_ECHILD;
-          NaClXCondVarBroadcast(&nap->children_cv);
-          NaClXMutexUnlock(&nap->children_mu);
-          goto out;
-        }
-        /* wait for next child to exit */
-        nap_child = DynArrayGet(&nap->children, cage_id);
-        if (nap_child) {
-          NaClLog(1, "Thread children count: %d\n", nap->num_children);
-          NaClXCondVarTimedWaitRelative(&nap->children_cv, &nap->children_mu, &timeout);
-          /* exit if selected child has finished */
-          if (!(nap_child = DynArrayGet(&nap->children, cage_id)) || !nap_child->running) {
-            ret = cage_id;
-            NaClXCondVarBroadcast(&nap->children_cv);
-            NaClXMutexUnlock(&nap->children_mu);
-            if (nap_child) *stat_loc_ptr = nap_child->exit_status;
-            else {
-              // this could be a zombie so lets check in this case
-              struct NaClZombie* zombie = NaClCheckZombies(nap);
-              *stat_loc_ptr = zombie->exit_status;
-            }
-            NaClRemoveZombie(nap, cage_id); //remove from zombie list regardless
-            goto out;
-          }
-        }
-    
-        NaClXCondVarBroadcast(&nap->children_cv);
+      // check the zombies dynarray, and lazily return the first exited process if it exists
+      struct NaClZombie* zombie = NaClCheckZombies(nap);
+      if (zombie) {
+        if (stat_loc_ptr) *stat_loc_ptr = zombie->exit_status; // set the status pointer if it exists
+        ret = zombie->cage_id;
+        NaClRemoveZombie(nap, ret); //remove from zombie list
 
+        NaClXCondVarBroadcast(&nap->children_cv);
         NaClXMutexUnlock(&nap->children_mu);
+
+        goto out;  
```

Comment:
You could potentially break instead of goto, but I don't think that's a big deal